### PR TITLE
enh: support reduction clause in openmp pass

### LINKED ISSUE #4136: variable increment example
Towards #3777

```fortran
subroutine increment_ctr(n, ctr)
use omp_lib
implicit none
integer, intent(in) :: n
integer, intent(inout) :: ctr

integer :: local_ctr

integer :: i

!$omp parallel private(i) reduction(+:local_ctr)
!$omp do
do i = 1, n
    local_ctr = local_ctr + 1
end do
!$omp end do
!$omp end parallel

ctr = ctr + local_ctr
end subroutine

program openmp
integer, parameter :: n = 1000000
integer :: ctr

ctr = 0
call increment_ctr(n, ctr)

print *, ctr
end program
```

Transformed to:

```fortran
module thread_data_module
    use, intrinsic :: iso_c_binding
    type, bind(C) :: thread_data
        integer(c_int) :: n, ctr
    end type thread_data
end module thread_data_module

subroutine lcompilers_increment_ctr(data) bind(C)
use thread_data_module
use iso_c_binding
use libgompinterop
implicit none
type(c_ptr), value :: data
type(thread_data), pointer :: tdata
integer(c_int) :: n, local_ctr

integer(c_int) :: i, num_threads, chunk, leftovers, thread_num, start, end

call c_f_pointer(data, tdata)

n = tdata%n

num_threads = omp_get_max_threads()
chunk = n / num_threads
leftovers = mod(n, num_threads)

thread_num = omp_get_thread_num()
start = chunk * thread_num

if (thread_num < leftovers) then
    start = start + thread_num
else
    start = start + leftovers
end if

end = start + chunk

if (thread_num < leftovers) then
    end = end + 1
end if

local_ctr = 0

do i = start + 1, end
    local_ctr = local_ctr + 1
end do

call GOMP_barrier()

tdata%ctr = tdata%ctr + local_ctr

end subroutine

subroutine increment_ctr(n, ctr)
use thread_data_module
use libgompinterop
implicit none

integer(c_int), intent(in) :: n
integer(c_int), intent(inout) :: ctr

type(thread_data), target :: data
type(c_ptr) :: tdata

interface
subroutine lcompilers_increment_ctr(data) bind(C)
use iso_c_binding
type(c_ptr), value :: data
end subroutine
end interface

data%n = n
data%ctr = ctr

tdata = c_loc(data)

call GOMP_parallel(c_funloc(lcompilers_increment_ctr), tdata, 0, 0)

ctr = data%ctr

end subroutine

program openmp
use libgompinterop
use thread_data_module
implicit none

integer(c_int) :: n = 1000000, ctr

ctr = 0
call increment_ctr(n, ctr)

print *, "Done"
print *, ctr

end program
```

```console
% lfortran -c libgompinterop.f90 && lfortran b-transformed.f90 -L/Users/pranavchiku/repos/llvm-project/openmp/build/runtime/src/ -lomp -Wl,-rpath,/Users/pranavchiku/repos/llvm-project/openmp/build/runtime/src/           
Done
1000000
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Closes #4136, towards #3777.

Now next step is to get array example working for which we fix https://github.com/lfortran/lfortran/issues/4031#issuecomment-2143981388

### CODE REVIEW

File: integration_tests/openmp_06.f90 [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+local_ctr = 0
+!$omp parallel private(i) reduction(+:local_ctr)
+!$omp do
+do i = 1, n
+    local_ctr = local_ctr + 1
+end do
+!$omp end do
+! the following condition shall be true almost always
+if (local_ctr > ( n / omp_get_max_threads()) + 5) error stop
```
Comment (Pranavchiku) [inline integration_tests/openmp_06.f90:19]: This ensures that we are not serially running the test, with the diff: 

```diff
diff --git a/integration_tests/openmp_06.f90 b/integration_tests/openmp_06.f90
index 5dba9b360..31456a226 100644
--- a/integration_tests/openmp_06.f90
+++ b/integration_tests/openmp_06.f90
@@ -16,7 +16,7 @@ do i = 1, n
 end do
 !$omp end do
 ! the following condition shall be true almost always
-if (local_ctr > ( n / omp_get_max_threads()) + 5) error stop
+if (local_ctr > ( n / 4) + 5) error stop
 !$omp end parallel
 
 ctr = ctr + local_ctr
@@ -27,7 +27,7 @@ use omp_lib
 integer, parameter :: n = 1000000
 integer :: ctr
 
-call omp_set_num_threads(8)
+! call omp_set_num_threads(8)
 ctr = 0
 call increment_ctr(n, ctr)
 
```

If we run it without `--openmp`, it stops.

```console
% lfortran ./integration_tests/openmp_06.f90 
ERROR STOP
% lfortran ./integration_tests/openmp_06.f90 --openmp --openmp-lib-dir=$CONDA_PREFIX/lib/
1000000
```

**Comment by Pranavchiku** [comment]
I think `CI / Check Release build (pull_request)` failure shall work on rerun.

**Comment by certik** [comment]
The failure is some warning in your code:
```
/home/runner/work/lfortran/lfortran/src/libasr/../libasr/asr.h:50367:27: error: ‘v.LCompilers::DoConcurrentStatementVisitor::<unnamed>.LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::DoConcurrentStatementVisitor>::current_expr’ may be used uninitialized [-Werror=maybe-uninitialized]
50367 |             ASR::expr_t** current_expr_copy_363 = current_expr;
      |                           ^~~~~~~~~~~~~~~~~~~~~
/home/runner/work/lfortran/lfortran/src/libasr/pass/openmp.cpp: In member function ‘LCompilers::ASR::symbol_t* LCompilers::DoConcurrentVisitor::create_lcompilers_function(const LCompilers::Location&, const LCompilers::ASR::DoConcurrentLoop_t&, std::map<std::__cxx11::basic_string<char>, LCompilers::ASR::ttype_t*>&, std::string)’:
/home/runner/work/lfortran/lfortran/src/libasr/pass/openmp.cpp:439:42: note: ‘v’ declared here
  439 |             DoConcurrentStatementVisitor v(al, current_scope);
      |                                          ^
cc1plus: all warnings being treated as errors
```
So that has to be fixed.

**Comment by Pranavchiku** [comment]
This is weird, if it was a warning then almost every test must fail at building, no? That's when I think it shall work at rerun because that part of code is unmodified.

**Review by certik — APPROVED** [review]
I think this looks great!

Does this also work with the native `do concurrent`? I think it might!

**Comment by Pranavchiku** [comment]
> Does this also work with the native do concurrent?

Only if we somehow mention that `local_ctr` is reduction variable.

**Comment by certik** [comment]
> This is weird, if it was a warning then almost every test must fail at building, no?

Every CI test that has `export CXXFLAGS="-Werror"` in it should fail indeed.

**Comment by certik** [comment]
Regarding `do concurrent`, here is a working example that runs today with LFortran serially:
```
subroutine increment_ctr(n, ctr)
implicit none
integer, intent(in) :: n
integer, intent(inout) :: ctr

integer :: local_ctr

integer :: i

local_ctr = 0
do concurrent (i = 1:n) reduce(+:local_ctr)
    local_ctr = local_ctr + 1
end do

ctr = ctr + local_ctr
end subroutine

program openmp_06
implicit none
integer, parameter :: n = 1000000
integer :: ctr

ctr = 0
call increment_ctr(n, ctr)

print *, ctr
if (ctr /= 1000000) error stop
end program
```

As you can see, it sets `reduce` correctly. All you need is to ensure it gets propagated to ASR, then everything should just work.

**Comment by Pranavchiku** [comment]
Perfect, that is what I was talking. Somehow parse `reduction` information to `ASR` and it should work. Maybe this can be `LFortran` extension or I am too naive to understand what can be put into it?

**Comment by certik** [comment]
The code in https://github.com/lfortran/lfortran/pull/4219#issuecomment-2159122608 should be standard Fortran 2023, so we have to support it, by generating the same ASR as this PR does for the OpenMP example. This should be very easy and you can do it in a subsequent PR.

**Comment by Pranavchiku** [comment]
I think CI should now be good.

**Pull request merged**